tests/longevity/longevity_test.go

```diff
@@ -0,0 +1,340 @@
+	})
+})
+
+func testTrigger(wg *sync.WaitGroup,
+	contexts []*scheduler.Context,
+	triggerType string,
+	triggerFunc func([]*scheduler.Context, *chan *EventRecord),
+	triggerLoc *sync.Mutex,
+	triggerEventsChan *chan *EventRecord) {
+	defer wg.Done()
+
+	minRunTime := Inst().MinRunTimeMins
+	timeout := (minRunTime) * 60
+
+	start := time.Now().Local()
+	lastInvocationTime := start
+
+	for {
+		// if timeout is 0, run indefinitely
+		if timeout != 0 && int(time.Since(start).Seconds()) > timeout {
+			break
+		}
+
+		// Get next interval of when trigger should happen
+		// This intercal can dynamically change by editing configMap
+		waitTime, err := getWaitTime(triggerType)
+		Expect(err).NotTo(HaveOccurred())
+
+		if waitTime != 0 {
+			// If trigger is not disabled
+			if time.Since(lastInvocationTime) > time.Duration(waitTime) {
+				// If its right time to trigger,
+				// check no other disruptive trigger is happening at same time
+				takeLock(triggerLoc, triggerType)
+
+				triggerFunc(contexts, triggerEventsChan)
+
+				if isDisruptiveTrigger(triggerType) {
+					triggerLoc.Unlock()
+				}
+
+				lastInvocationTime = time.Now().Local()
+			}
+		}
+		time.Sleep(time.Second * 15)
+	}
+}
+
+func watchConfigMap() error {
+	chaosMap = map[string]int{}
+	cm, err := core.Instance().GetConfigMap(testTriggersConfigMap, configMapNS)
+	if err != nil {
+		return fmt.Errorf("Error reading config map: %v", err)
+	}
+	err = populateDataFromConfigMap(&cm.Data)
+	if err != nil {
+		return err
+	}
+
+	// Apply watch if configMap exists
+	fn := func(object runtime.Object) error {
+		cm, ok := object.(*v1.ConfigMap)
+		if !ok {
+			err := fmt.Errorf("invalid object type on configmap watch: %v", object)
+			return err
+		}
+		if len(cm.Data) > 0 {
+			err = populateDataFromConfigMap(&cm.Data)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	err = core.Instance().WatchConfigMap(cm, fn)
+	if err != nil {
+		return fmt.Errorf("Failed to watch on config map: %s due to: %v", testTriggersConfigMap, err)
+	}
+	return nil
+}
+
+func populateDisruptiveTriggers() {
+	disruptiveTriggers = map[string]bool{
+		RestartVolDriver: false,
+		CrashVolDriver:   false,
+		RebootNode:       true,
+		EmailReporter:    false,
+		AppTaskDown:      false,
+	}
+}
+
+func isDisruptiveTrigger(triggerType string) bool {
+	return disruptiveTriggers[triggerType]
+}
+
+// takeLock takes lock only if no other disruptive trigger is happening at the same time.
+func takeLock(triggerLoc *sync.Mutex, triggerType string) {
+	if isDisruptiveTrigger(triggerType) {
+		// At a give point in time, only single disruptive trigger is allowed to run.
+		// No other disruptive or non-disruptive trigger can run at this time.
+		triggerLoc.Lock()
+	} else {
+		// If trigger is non-disruptive then just check if no other disruptive trigger is running or not
+		// and release the lock immidiately so that other non-disruptive triggers can happen.
+		triggerLoc.Lock()
+		triggerLoc.Unlock()
```

Comment:
remove

Reply:
@harsh-px this is required, for non-disruptive triggers, they need to check if any other disruptive triggers are happening or not. If it is able to acquire the lock here that means there no other disruptive event happening at this time. Then it can release the lock and continue ahead.